src/dtype.rs

```diff
@@ -244,9 +246,30 @@ impl_element_scalar!(Complex64 => NPY_CDOUBLE,
 #[cfg(any(target_pointer_width = "32", target_pointer_width = "64"))]
 impl_element_scalar!(usize, isize);
 
-unsafe impl Element for PyObject {
+unsafe impl<T> Element for Py<T>
+where
+    T: PyTypeInfo + 'static,
+{
     const IS_COPY: bool = false;
 
+    fn check_element_types<D: Dimension>(py: Python, array: &PyArray<Self, D>) -> PyResult<()> {
```

Comment:
What I don't like is that there's again logic being added into `Element`, so it goes from being a descriptive trait to an 'actionable' trait.

Perhaps something like this might be cleaner?

```rust
pub unsafe trait Element {
    const IS_COPY: bool;
    fn py_type(py: Python) -> Option<Self>; // None for anything with IS_COPY=true or PyAny
    fn get_dtype(py: Python) -> &PyArrayDescr;
}
```

Reply:
> What I personally don't like is that there's again logic being added into Element, so it goes from being a descriptive trait to an 'actionable' trait once again.

The point of this was to make the policy changeable by user code, i.e. I really known I'll get `PyArray<T, D>` and do not want to pay the cost of checking the types, I can implement this unsoundly myself. But I guess due to the general issues, using `PyArray<PyObject, D>` together with `unchecked_downcast` would be the way to make this work. We just need to document this...

Reply:
I guess my point was, it doesn't feel like it should be stuffed in here. Do you have an example in mind where you would implement custom that *doesn't* just go through the elements and checks dtypes?

Another disadantage here being, if I *do* want to do something custom downstream but similar to what's being done here, I would basically have to literally copy-paste rust-numpy source code that iterates over numpy array items - that's not very nice.

And finally, the "don't check this" part doesn't sound like it belongs to the type itself, but rather to the call point. If you bind it to the type, you can't do "here I know it's safe, don't check, but here it's arbitrary user input, need to check" which would be a fairly reasonable use case, I believe. This you could probably achieve by having a separate trait on `PyArray<T, D>` (pulling this stuff out of `T: Element`) with default impl that always checks instance types for any `Py<T>` that's not `PyAny`, and then add a newtype wrapper like `Unchecked<PyArray<T, D>>` that would override this and impl a `FromPyObject` for it. Just a thought.